### PR TITLE
Initial Implementation

### PR DESCRIPTION
**Pull request by blink1073**
For discussion.

### CODE REVIEW

File: plugins/menus/src/constraints.ts [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+
+/**
+ * The interface an object is required to implement for a Constraint.
+ *
+ * These are very simple constraint systems which are designed to 
+ * produce Before/After relationships in the form of graph edges
+ * for input into a solver.
+ */
+export
+interface IConstraint {
```
Comment (dwillmer) [inline plugins/menus/src/constraints.ts]: we're not using IConstraint or the Before/After classes anymore, so this whole file can be removed.

File: plugins/menus/src/index.ts [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+} from 'phosphor-signaling';
+
+import {
+  attachWidget, detachWidget
+} from 'phosphor-widget';
+
+
+import './index.css';
+
+export * from './constraints';
```
Comment (dwillmer) [inline plugins/menus/src/index.ts]: ...and this export can be removed, due to the IConstraint removal.

**Comment by dwillmer** [comment]
WRT the 2 comments above - you're using an older version of the menu solver code, but everything looks good to me. i suggest merging and updating later.

**Comment by dwillmer** [comment]
LGTM, +1 for merging.

**Comment by blink1073** [comment]
Closing in favor of #3.

**Pull request closed without merging**